### PR TITLE
attach-to-chat 2: connect

### PR DESCRIPTION
**Pull request by songgao**
Not triggering any RPC yet, but components are connected. Navigation, filtering, etc. work. I'll have a follow-up PR for actually making those RPCs and likely another to stream from SimpleFS to chat.

### CODE REVIEW

**Comment by songgao** [comment]
bump!

File: shared/chat/conversation-list/conversation-list-container.js [hunk 1]
```diff
@@ -1,15 +1,133 @@
...
+const boolToNumber = (b: boolean): number => (b ? 1 : 0)
+
+const staleToNumber = (convTime: number, staleCutoff: number) => (convTime < staleCutoff ? 1 : 0)
+
+const getAWeekAgo = () => {
+  let t = new Date()
+  return t.setDate(t.getDate() - 7) // works fine for cross-boundary; returns a number
+}
+
+const getSortedConversationIDKeys = memoize(
```
Comment (chrisnojima) [inline shared/chat/conversation-list/conversation-list-container.js]: really all this should move to core, and will soon

File: shared/reducers/fs.js [hunk 2]
```diff
@@ -292,6 +297,13 @@ export default function(state: Types.State = initialState, action: FsGen.Actions
     case FsGen.setSendLinkToChatChannels:
       // $FlowIssue
       return state.setIn(['sendLinkToChat', 'channels'], action.payload.channels)
+    case FsGen.setSendAttachmentToChatConvID:
+      // $FlowIssue
```
Comment (chrisnojima) [inline shared/reducers/fs.js]: why flow issue

Comment (songgao) [reply shared/reducers/fs.js]: I think `setIn` with flow is just broken. I'll see if I can get around it with other mutaters.

File: shared/chat/conversation-list/conversation-list-container.js (hunk 1 above)
Comment (songgao) [reply shared/chat/conversation-list/conversation-list-container.js]: Cool; keep this for now? Also note that this is different from inbox sorting.

File: shared/actions/fs/index.js [hunk 3]
```diff
@@ -899,6 +911,10 @@ function* fsSaga(): Saga.SagaGenerator<any, any> {
   yield* Saga.chainAction<FsGen.ShowMoveOrCopyPayload>(FsGen.showMoveOrCopy, showMoveOrCopy)
   yield* Saga.chainAction<FsGen.CancelMoveOrCopyPayload>(FsGen.cancelMoveOrCopy, cancelMoveOrCopy)
   yield* Saga.chainGenerator<FsGen.ShowSendLinkToChatPayload>(FsGen.showSendLinkToChat, showSendLinkToChat)
+  yield* Saga.chainGenerator<FsGen.ShowSendLinkToChatPayload>(
```
Comment (chrisnojima) [inline shared/actions/fs/index.js]: should just be chainAction

File: shared/chat/conversation-list/conversation-list.js [hunk 4]
```diff
@@ -81,8 +83,19 @@ const ConversationList = (props: Props) => {
...
-      {!!props.filter && <ConversationFilterInput style={styles.filter} {...props.filter} />}
+    <Kb.Box2 direction="vertical" fullWidth={true} fullHeight={true}>
+      {!!props.filter && (
+        <ConversationFilterInput
+          style={styles.filter}
+          isLoading={props.filter.isLoading}
+          filter={props.filter.filter}
+          onSetFilter={props.filter.onSetFilter}
+          filterFocusCount={0}
+          onSelectDown={() => {} /* TODO: make these work */}
```
Comment (chrisnojima) [inline shared/chat/conversation-list/conversation-list.js]: where is this implemented

**Comment by songgao** [comment]
I'm gonna squash all these commit into one so conflicts don't kill me

**Comment by songgao** [comment]
ready for another look!

**Comment by songgao** [comment]
⪮

**Comment by songgao** [comment]
Actually making attachment happen was simpler than I thought, and since this hasn't been approved yet I just squeezed it into this branch instead of creating a new one.

**Comment by songgao** [comment]
rebased and bump!

**Comment by songgao** [comment]
bump!

File: shared/chat/conversation-list/choose-conversation.js [hunk 5]
```diff
@@ -22,27 +23,31 @@ class ChooseConversation extends React.Component<Props & Kb.OverlayParentProps,
...
-    this.setState({expanded: false})
-  }
+  _onDone = () => this.setState({expanded: false})
   render() {
     return (
       <>
         <Kb.DropdownButton
-          disabled={this.props.disabled}
-          selected={<Kb.Text type="BodySemibold">{this.props.dropdownButtonDefaultText}</Kb.Text>}
+          disabled={false}
```
Comment (chrisnojima) [inline shared/chat/conversation-list/choose-conversation.js]: 🔥

File: shared/chat/conversation-list/choose-conversation.js [hunk 6]
```diff
@@ -22,27 +23,31 @@ class ChooseConversation extends React.Component<Props & Kb.OverlayParentProps,
...
           setAttachmentRef={this.props.setAttachmentRef}
           toggleOpen={this._toggleOpen}
           style={Styles.collapseStyles([styles.dropdownButton, this.props.dropdownButtonStyle])}
         />
         <Kb.Overlay
           attachTo={this.props.getAttachmentRef}
           onHidden={this._toggleOpen}
-          position={this.props.position || 'center center'}
-          style={Styles.collapseStyles([styles.overlay, this.props.overlayStyle])}
+          position={'center center'}
```
Comment (chrisnojima) [inline shared/chat/conversation-list/choose-conversation.js]: dont need curlies

File: shared/chat/conversation-list/conversation-list-container.js [hunk 7]
```diff
@@ -1,17 +1,189 @@
...
+    return metaMap
+      .valueSeq()
+      .toArray()
+      .sort((a, b) => {
+        // leveled order rules:
+        // 1. unmuted before muted
+        // 2. active conversations before inactive (has activity in the past week)
+        // 3. notification type: onAnyActivity before onWhenAtMentioned, before never
+        // 4. activity timestamp being the last tie breaker
+        const mutedBased = boolToNumber(a.isMuted) - boolToNumber(b.isMuted)
```
Comment (chrisnojima) [inline shared/chat/conversation-list/conversation-list-container.js]: if the goal is to see if they're both the same truthy vs falsey you can just be
```
const muteBased = !a.isMuted !== !b.isMuted
```

File: shared/chat/conversation-list/conversation-list.js [hunk 8]
```diff
@@ -81,8 +87,23 @@ const ConversationList = (props: Props) => {
...
+      {!!props.filter && (
+        <ConversationFilterInput
+          onBack={props.onBack}
+          style={styles.filter}
+          isLoading={props.filter.isLoading}
+          filter={props.filter.filter}
+          onSetFilter={props.filter.onSetFilter}
+          filterFocusCount={0}
+          noShortcut={true}
+          onBlur={() => {}}
```
Comment (chrisnojima) [inline shared/chat/conversation-list/conversation-list.js]: is it safe to just have these do nothing now? why?

File: shared/chat/conversation-list/conversation-list-container.js (hunk 7 above)
Comment (songgao) [reply shared/chat/conversation-list/conversation-list-container.js]: I still need a number here to return though, so would still need to use boolToNumber.

File: shared/chat/conversation-list/choose-conversation.js (hunk 5 above)
Comment (songgao) [reply shared/chat/conversation-list/choose-conversation.js]: You mean we should keep this prop? This is only used in the attachment popup. I think it's always enabled.

Comment (chrisnojima) [reply shared/chat/conversation-list/choose-conversation.js]: no i mean you can just delete it, it defaults to false right

File: shared/chat/conversation-list/conversation-list.js (hunk 8 above)
Comment (songgao) [reply shared/chat/conversation-list/conversation-list.js]: I actually just didn't know what they were for, and this isn't used in chat at all, just in the attachment popup I'm building now, so I left them like that. But I guess I'll just figure out and do them right here.

File: shared/chat/conversation-list/choose-conversation.js (hunk 5 above)
Comment (songgao) [reply shared/chat/conversation-list/choose-conversation.js]: Ah that's right!

File: shared/chat/conversation-list/conversation-list.js (hunk 8 above)
Comment (songgao) [reply shared/chat/conversation-list/conversation-list.js]: OK I don't think we need these for choose-conversation, but I added focusOnMount for the popup.

**Comment by songgao** [comment]
bump!

**Comment by songgao** [comment]
⪮

**Comment by songgao** [comment]
⪮

**Comment by songgao** [comment]
woohooo thanks Nojima

**Pull request merged**